src/classes/HiFiCommunicator.ts

```diff
@@ -305,6 +406,78 @@ export class HiFiCommunicator {
+     * and failures and the like). This method is passed into the MixerSession to serve as its state change handler.
+     * @param state
+     */
+    private _stateChangeCoordinator(state: HiFiConnectionStates): void {
+        /**
+         TODO: This (and various other functions, including setting of timeouts and the
+         main connection method) need to do the following:
+
+         A) When _first_ attempting a connection, the main connection method should set
+            this._currentHiFiConnectionState = HiFiConnectionStates.Connecting
+            before doing anything else.
+
+         B) When handling a state change, this method (and/or methods it calls) should do the following:
+
+            i)    IF the state change got triggered via an explicit call to disconnectFromHiFiAudioAPIServer()
+                  THEN handle the new state as implemented below
+                  (note: this behavior might be most easily implemented through the use of
+                  an explicit, transient, DISCONNECTING state.)
+
+            ii) IF this._currentHiFiConnectionState is currently CONNECTING:
+
+               1. IF the new state is DISCONNECTED, FAILED, or UNAVAILABLE
+                  AND IF autoRetryInitialConnection is true
+                  AND IF maxSecondsToSpendRetryingInitialConnection has not yet elapsed
+                  THEN ensure that this._currentHiFiConnectionState stays at CONNECTING
+                  ELSE handle the new state as implemented below
+
+               2. ELSE IF the new state is CONNECTED
+                  THEN handle the new state as implemented below
+                  AND kill the maxSecondsToSpendRetryingInitialConnection timeout if it's set
+
+               3. ELSE IF the new state is other (CONNECTING or RECONNECTING)
+                  AND IF the new state is not the same as this._currentHiFiConnectionState
+                  THEN handle the new state as implemented below
+
+            ii) IF this._currentHiFiConnectionState is currently CONNECTED or RECONNECTING:
+
+               1. IF the new state is DISCONNECTED, FAILED, or UNAVAILABLE
+                  AND IF autoRetryOnDisconnect is true
+                  AND IF maxSecondsToSpendRetryingOnDisconnect has not yet elapsed
+                  THEN set this._currentHiFiConnectionState to RECONNECTING
+                  AND try to reconnect
+                  ELSE handle the new state as implemented below
+
+               2. ELSE IF the new state is CONNECTED
+                  THEN handle the new state as implemented below
+                  AND kill the maxSecondsToSpendRetryingOnDisconnect timeout if it's set
+
+               4. ELSE IF the new state is other (CONNECTING or RECONNECTING)
+                  AND IF the new state is not the same as this._currentHiFiConnectionState
+                  THEN handle the new state as implemented below
```

Comment:
There are complex bits of external state that interact with all of this. My instinct is to NOT try to work through these now, but I'll raise it to give folks a chance to argue otherwise:
- How does this interact with being offline? Is that the same as our server reporting "unavailable" for capacity? Should we set up handler to "try again" when one comes back online, or is that for the app to decide (because it may involve asking the user if that's what they want), or does the answer depend on how long we were off line, or on what our _currentHiFiConnectionState was when we went offline?
- Same question for the page not being visible, and is that different for desktop v mobile.

Reply:
Yeah, I agree with your instinct on this! To answer the direct question, if the user goes offline it will manifest as a dropped connection first, and then an "unable to make signaling connection" error for any retries, because the WebSocket attempt will immediately fail.

But I think that we should not set up handlers to watch connection state and make decisions based on that -- instead, simply retry as instructed for the amount of time instructed, and if it still can't connect just move to a failed state and stop trying.